File: Unit-1/CheckTheCheck.cpp

// Unit = 1
// Program Name = Check The Check

// Name  = Varun Kotgire
// class = TYCSE-A
// Roll  = 79
// Batch = 4

#include <vector>
#include <iostream>
#include <string>

using namespace std;

bool handleStraight(const vector<string> &board, int x, int y, char target1, char target2)
{
    if ((x - 1 >= 0 && (board[y][x - 1] == target1 || board[y][x - 1] == target2)) ||
        (x + 1 < 8 && (board[y][x + 1] == target1 || board[y][x + 1] == target2)) ||
        (y - 1 >= 0 && (board[y - 1][x] == target1 || board[y - 1][x] == target2)) ||
        (y + 1 < 8 && (board[y + 1][x] == target1 || board[y + 1][x] == target2)))
        return true;

    for (int xc = -1; xc + x - 1 >= 0 && board[y][xc + x] == '.'; --xc)
        if (board[y][xc + x - 1] == target1 || board[y][xc + x - 1] == target2)
            return true;

    for (int xc = 1; xc + x + 1 < 8 && board[y][xc + x] == '.'; ++xc)
        if (board[y][xc + x + 1] == target1 || board[y][xc + x + 1] == target2)
            return true;

    for (int yc = -1; yc + y - 1 >= 0 && board[y + yc][x] == '.'; --yc)
        if (board[y + yc - 1][x] == target1 || board[y + yc - 1][x] == target2)
            return true;

    for (int yc = 1; yc + y + 1 < 8 && board[y + yc][x] == '.'; ++yc)
        if (board[y + yc + 1][x] == target1 || board[y + yc + 1][x] == target2)
            return true;

    return false;
}

bool handleDiagonal(const vector<string> &board, int x, int y, char target1, char target2)
{
    bool upLeft(true), upRight(true), downLeft(true), downRight(true);

    for (int c(1); upLeft || upRight || downLeft || downRight; ++c)
    {
        if (upLeft)
        {
            if (x - c >= 0 && y - c >= 0)
            {
                if (board[y - c][x - c] == target1 || board[y - c][x - c] == target2)
                    return true;
                else if (board[y - c][x - c] != '.')
                    upLeft = false;
            }
            else
            {
                upLeft = false;
            }
        }

        if (upRight)
        {
            if (x + c < 8 && y - c >= 0)
            {
                if (board[y - c][x + c] == target1 || board[y - c][x + c] == target2)
                    return true;
                else if (board[y - c][x + c] != '.')
                    upRight = false;
            }
            else
            {
                upRight = false;
            }
        }

        if (downLeft)
        {
            if (x - c >= 0 && y + c < 8)
            {
                if (board[y + c][x - c] == target1 || board[y + c][x - c] == target2)
                    return true;
                else if (board[y + c][x - c] != '.')
                    downLeft = false;
            }
            else
            {
                downLeft = false;
            }
        }

        if (downRight)
        {
            if (x + c < 8 && y + c < 8)
            {
                if (board[y + c][x + c] == target1 || board[y + c][x + c] == target2)
                    return true;
                else if (board[y + c][x + c] != '.')
                    downRight = false;
            }
            else
            {
                downRight = false;
            }
        }
    }

    return false;
}

bool handleKnight(const vector<string> &board, int x, int y, char target)
{
    int knightMoves[8][2] = {{-2, -1}, {-2, 1}, {2, -1}, {2, 1}, {-1, -2}, {1, -2}, {-1, 2}, {1, 2}};
    for (int i = 0; i < 8; ++i)
    {
        int newX = x + knightMoves[i][0];
        int newY = y + knightMoves[i][1];
        if (newX >= 0 && newX < 8 && newY >= 0 && newY < 8 && board[newY][newX] == target)
        {
            return true;
        }
    }
    return false;
}

bool handlePawn(const vector<string> &board, int x, int y, char target, int yChange)
{
    if (y + yChange < 8 && y + yChange >= 0)
    {
        if (x - 1 >= 0 && board[y + yChange][x - 1] == target)
            return true;
        if (x + 1 < 8 && board[y + yChange][x + 1] == target)
            return true;
    }
    return false;
}

int main()
{
    bool forever = true;
    int whitePosx, blackPosx, whitePosy, blackPosy, temp;
    vector<string> board(8);

    for (int t = 1; forever; ++t)
    {
        whitePosx = blackPosx = -1;
        for (int i = 0; i < 8; ++i)
        {
            cin >> board[i];
            temp = board[i].find('k');
            if (temp != string::npos)
            {
                blackPosy = i;
                blackPosx = temp;
            }
            temp = board[i].find('K');
            if (temp != string::npos)
            {
                whitePosy = i;
                whitePosx = temp;
            }
        }

        if (whitePosx == -1)
            break;

        cout << "Game #" << t << ": ";

        if (handlePawn(board, whitePosx, whitePosy, 'p', -1))
            cout << "white king is in check.\n";
        else if (handlePawn(board, blackPosx, blackPosy, 'P', 1))
            cout << "black king is in check.\n";
        else if (handleKnight(board, whitePosx, whitePosy, 'n'))
            cout << "white king is in check.\n";
        else if (handleKnight(board, blackPosx, blackPosy, 'N'))
            cout << "black king is in check.\n";
        else if (handleStraight(board, whitePosx, whitePosy, 'r', 'q'))
            cout << "white king is in check.\n";
        else if (handleStraight(board, blackPosx, blackPosy, 'R', 'Q'))
            cout << "black king is in check.\n";
        else if (handleDiagonal(board, whitePosx, whitePosy, 'b', 'q'))
            cout << "white king is in check.\n";
        else if (handleDiagonal(board, blackPosx, blackPosy, 'B', 'Q'))
            cout << "black king is in check.\n";
        else
            cout << "no king is in check.\n";
    }
}

// Sample Input

// ..k.....
// ppp.pppp
// ........
// .R...B..
// ........
// ........
// PPPPPPPP
// K.......
// rnbqkbnr
// pppppppp
// ........
// ........
// ........
// ........
// PPPPPPPP
// RNBQKBNR
// rnbqk.nr
// ppp..ppp
// ....p...
// ...p....
// .bPP....
// .....N..
// PP..PPPP
// RNBQKB.R
// ........
// ........
// ........
// ........
// ........
// ........
// ........
// ........

// Sample Output

// Game #1: black king is in check.
// Game #2: no king is in check.
// Game #3: white king is in check.

File: Unit-1/CheckTheCheck.py
def handle_straight(board, x, y, target1, target2):
    if (x - 1 >= 0 and (board[y][x - 1] == target1 or board[y][x - 1] == target2)) or \
       (x + 1 < 8 and (board[y][x + 1] == target1 or board[y][x + 1] == target2)) or \
       (y - 1 >= 0 and (board[y - 1][x] == target1 or board[y - 1][x] == target2)) or \
       (y + 1 < 8 and (board[y + 1][x] == target1 or board[y + 1][x] == target2)):
        return True

    for xc in range(-1, -x - 1, -1):
        if board[y][xc + x] != '.':
            if board[y][xc + x] == target1 or board[y][xc + x] == target2:
                return True
            break

    for xc in range(1, 8 - x):
        if board[y][xc + x] != '.':
            if board[y][xc + x] == target1 or board[y][xc + x] == target2:
                return True
            break

    for yc in range(-1, -y - 1, -1):
        if board[yc + y][x] != '.':
            if board[yc + y][x] == target1 or board[yc + y][x] == target2:
                return True
            break

    for yc in range(1, 8 - y):
        if board[yc + y][x] != '.':
            if board[yc + y][x] == target1 or board[yc + y][x] == target2:
                return True
            break

    return False


def handle_diagonal(board, x, y, target1, target2):
    up_left = up_right = down_left = down_right = True

    for c in range(1, 8):
        if up_left:
            if x - c >= 0 and y - c >= 0:
                if board[y - c][x - c] == target1 or board[y - c][x - c] == target2:
                    return True
                elif board[y - c][x - c] != '.':
                    up_left = False
            else:
                up_left = False

        if up_right:
            if x + c < 8 and y - c >= 0:
                if board[y - c][x + c] == target1 or board[y - c][x + c] == target2:
                    return True
                elif board[y - c][x + c] != '.':
                    up_right = False
            else:
                up_right = False

        if down_left:
            if x - c >= 0 and y + c < 8:
                if board[y + c][x - c] == target1 or board[y + c][x - c] == target2:
                    return True
                elif board[y + c][x - c] != '.':
                    down_left = False
            else:
                down_left = False

        if down_right:
            if x + c < 8 and y + c < 8:
                if board[y + c][x + c] == target1 or board[y + c][x + c] == target2:
                    return True
                elif board[y + c][x + c] != '.':
                    down_right = False
            else:
                down_right = False

    return False


def handle_knight(board, x, y, target):
    knight_moves = [(-2, -1), (-2, 1), (2, -1), (2, 1), (-1, -2), (1, -2), (-1, 2), (1, 2)]
    for move in knight_moves:
        new_x, new_y = x + move[0], y + move[1]
        if 0 <= new_x < 8 and 0 <= new_y < 8 and board[new_y][new_x] == target:
            return True
    return False


def handle_pawn(board, x, y, target, y_change):
    if 0 <= y + y_change < 8:
        if x - 1 >= 0 and board[y + y_change][x - 1] == target:
            return True
        if x + 1 < 8 and board[y + y_change][x + 1] == target:
            return True
    return False


def main():
    forever = True
    t = 1

    while forever:
        board = [input().strip() for _ in range(8)]
        input()  # consume the blank line

        white_posx = black_posx = -1

        for i in range(8):
            temp = board[i].find('k')
            if temp != -1:
                black_posy = i
                black_posx = temp
            temp = board[i].find('K')
            if temp != -1:
                white_posy = i
                white_posx = temp

        if white_posx == -1:
            break

        print(f"Game #{t}: ", end="")
        t += 1

        if handle_pawn(board, white_posx, white_posy, 'p', -1):
            print("white king is in check.")
        elif handle_pawn(board, black_posx, black_posy, 'P', 1):
            print("black king is in check.")
        elif handle_knight(board, white_posx, white_posy, 'n'):
            print("white king is in check.")
        elif handle_knight(board, black_posx, black_posy, 'N'):
            print("black king is in check.")
        elif handle_straight(board, white_posx, white_posy, 'r', 'q'):
            print("white king is in check.")
        elif handle_straight(board, black_posx, black_posy, 'R', 'Q'):
            print("black king is in check.")
        elif handle_diagonal(board, white_posx, white_posy, 'b', 'q'):
            print("white king is in check.")
        elif handle_diagonal(board, black_posx, black_posy, 'B', 'Q'):
            print("black king is in check.")
        else:
            print("no king is in check.")


if __name__ == "__main__":
    main()
